src/main/java/com/amazon/ion/impl/Ion11Writer.kt

```diff
@@ -0,0 +1,153 @@
+package com.amazon.ion.impl
+
+import com.amazon.ion.Decimal
+import com.amazon.ion.IonType
+import com.amazon.ion.Timestamp
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.time.Instant
+import java.time.LocalDate
+
+/**
+ * Writes Ion 1.1 data to an output source.
+ *
+ * This interface allows the user to write Ion data without being concerned about which output format is being used.
+ */
+interface Ion11Writer {
+
+    /**
+     * Indicates that writing is completed and all buffered data should be written and flushed as if this were the end
+     * of the Ion data stream. For example, an Ion binary writer will finalize any local symbol table, write all
+     * top-level values, and then flush.
+     *
+     * This method may only be called when all top-level values are completely written and (`stepped out`)[stepOut].
+     *
+     * Implementations should allow the application to continue writing further top-level values following the semantics
+     * for concatenating Ion data streams.
+     */
+    fun finish()
+
+    /**
+     * Closes this stream and releases any system resources associated with it.
+     * If the stream is already closed then invoking this method has no effect.
+     *
+     * If the cursor is between top-level values, this method will [finish] before closing the underlying output stream.
+     * If not, the resulting data may be incomplete and invalid Ion.
+     */
+    fun close()
+
+    /** Returns true if the writer is currently in a struct (indicating that field names are required). */
+    fun isInStruct(): Boolean
+
+    /** Returns the current depth of containers the writer is at.  This is 0 if the writer is at top-level. */
+    fun depth(): Int
+
+    /**
+     * Writes the Ion 1.1 IVM. IVMs can only be written at the top level of an Ion stream.
+     * @throws com.amazon.ion.IonException if in any container.
+     */
+    fun writeIVM()
+
+    /** Writes one annotation for the next value. */
+    fun writeAnnotations(annotation0: Long)
+
+    /** Writes two annotations for the next value. */
+    fun writeAnnotations(annotation0: Long, annotation1: Long)
+
+    /** Writes three or more annotations for the next value. */
+    fun writeAnnotations(annotation0: Long, annotation1: Long, vararg annotations: Long)
```

Comment:
Do we need variants that allow mixing Strings and Longs, for use with the FlexSym encoding?

Reply:
That's a good point. I'll add overloads that accept `SymbolToken`s.

Reply:
`SymbolToken` isn't ideal because it may require allocating a new one just to specify that we want text and not a symbol ID, or vice versa. Is there some way to finesse this in Kotlin? Otherwise, maybe the API is `addAnnotation` with variants for both String and Long...

Reply:
I could create inline value classes.
```kotlin
@JvmInline value class SymbolText(val text: String): SymbolToken {
    override fun getText(): String = text
    override fun assumeText(): String = text
    override fun getSid(): Int = 0
}
@JvmInline value class SymbolId(val id: Int): SymbolToken {
    override fun getText(): String? = null
    override fun assumeText(): Nothing = throw UnknownSymbolException(id)
    override fun getSid(): Int = id
}
```

If that's not okay, I can change to something like this.
```kotlin
fun addAnnotation(annotation: CharSequence)
fun addAnnotation(annotation: Long)
fun writeAnnotations()
```

Reply:
Yeah, the inline value classes look like a good start. Let's go with that for now. (Note: the SID for `SymbolText` should be `-1` (there's an `UNKNOWN_SID = -1` constant somewhere), since 0 is a valid SID.

Reply:
Actually, the value classes might not play nice with Java. I'll have to double check that.

Reply:
I ended up just adjusting the contract for `writeAnnotations` so that it is legal to call more than once preceding a value. I don't want to have a mix of `addAnnotations` and `writeAnnotations` because it would bloat the number of APIs for writing annotations and because it could be confusing having two different ways of writing annotations with different semantics.